File: snippets/design-patterns/behavioral/chain_of_responsibility.py
'''
--- Chain of Responsibility Pattern ---
Use Case: Pass a request along a chain of handlers until one processes it.
Example: Logging pipelines, HTTP middleware, or validation chains.
'''

class Handler:
    def __init__(self, successor=None):
        self.successor = successor

    def handle(self, request):
        if self.successor:
            return self.successor.handle(request)
        return None

class ConcreteHandlerA(Handler):
    def handle(self, request):
        if request == "A":
            return f"Handler A processed {request}"
        return super().handle(request)

class ConcreteHandlerB(Handler):
    def handle(self, request):
        if request == "B":
            return f"Handler B processed {request}"
        return super().handle(request)

# Example Usage
handler_chain = ConcreteHandlerA(ConcreteHandlerB())

print(handler_chain.handle("A"))  # Output: Handler A processed A
print(handler_chain.handle("B"))  # Output: Handler B processed B
print(handler_chain.handle("C"))  # Output: None

File: snippets/design-patterns/behavioral/chain_of_responsibility.go
// --- Chain of Responsibility Pattern ---
// Use Case: Pass a request along a chain of handlers until one processes it.
// Example: Logging pipelines, HTTP middleware, or validation chains.
// Description: The Chain of Responsibility Pattern decouples the sender of a request
// from its receiver, allowing multiple handlers to process the request dynamically.

package main

import "fmt"

type Handler interface {
	SetNext(handler Handler) Handler
	Handle(request string)
}

type BaseHandler struct {
	next Handler
}

func (h *BaseHandler) SetNext(handler Handler) Handler {
	h.next = handler
	return handler
}

func (h *BaseHandler) Handle(request string) {
	if h.next != nil {
		h.next.Handle(request)
	}
}

type ConcreteHandlerA struct {
	BaseHandler
}

func (h *ConcreteHandlerA) Handle(request string) {
	if request == "A" {
		fmt.Println("Handler A processed the request")
	} else {
		h.BaseHandler.Handle(request)
	}
}

type ConcreteHandlerB struct {
	BaseHandler
}

func (h *ConcreteHandlerB) Handle(request string) {
	if request == "B" {
		fmt.Println("Handler B processed the request")
	} else {
		h.BaseHandler.Handle(request)
	}
}

func main() {
	handlerA := &ConcreteHandlerA{}
	handlerB := &ConcreteHandlerB{}
	handlerA.SetNext(handlerB)

	handlerA.Handle("A") // Output: Handler A processed the request
	handlerA.Handle("B") // Output: Handler B processed the request
	handlerA.Handle("C") // No output
}
